File: backend/src/services/auditService.js
// src/services/auditService.js
// Trilho de auditoria financeira — APPEND-ONLY. Este serviço só INSERE; não
// existe (de propósito) nenhum caminho de update/delete de registos.
//
// Uso:
//   const actor = auditService.actorFrom(req);
//   await auditService.record(prismaOuTx, { actor, action, entityType, ... });
//
// Nas operações de dinheiro (pagamento) o registo é escrito DENTRO da mesma
// transação — ou fica tudo, ou nada. Nas restantes, é escrito logo após o
// sucesso; se falhar, a operação de negócio NÃO é revertida, mas o erro é
// registado com peso (logger + Sentry) para nunca passar despercebido.
const prisma = require('../config/database');
const logger = require('../config/logger');
const { captureException } = require('../config/sentry');

// Snapshot do ator a partir do pedido autenticado (desnormalizado de propósito).
function actorFrom(req) {
  return {
    actorId: req.user?.id ?? null,
    actorName: req.user?.name ?? null,
    actorRole: req.user?.role ?? null,
    companyId: req.user?.companyId ?? null,
    ip: req.ip ?? null,
  };
}

// Ator de um pedido AINDA NÃO autenticado (tentativa de login, recuperação de
// senha). Não há utilizador — o que identifica a tentativa é a origem, por isso
// vai o IP e o agente. Sem isto, uma sequência de tentativas falhadas não deixa
// rasto nenhum.
function anonimoFrom(req) {
  return {
    actorId: null,
    actorName: null,
    actorRole: null,
    companyId: null,
    ip: req.ip ?? null,
  };
}

// Contexto do pedido para o campo `detail`: user agent truncado (é livre e vem
// do cliente) e origem.
function contextoFrom(req) {
  return {
    ip: req.ip ?? null,
    agente: String(req.get?.('user-agent') || '').slice(0, 180) || null,
  };
}

/**
 * Escreve um registo de auditoria.
 * @param client prisma ou tx (para escrever dentro de uma transação)
 * @param entry  { actor, action, entityType, entityId?, entityRef?, detail? }
 */
async function record(client, { actor = {}, action, entityType, entityId = null, entityRef = null, detail = null }) {
  return client.auditLog.create({
    data: { action, entityType, entityId, entityRef, detail, ...actor },
  });
}

// Variante "não pode partir o negócio": regista após o sucesso da operação;
// uma falha aqui é grave (fica um buraco no trilho) — vai para o log e Sentry.
async function recordSafe(entry) {
  try {
    await record(prisma, entry);
  } catch (err) {
    logger.error('AUDITORIA FALHOU — registo perdido', { action: entry.action, message: err.message });
    captureException(err);
  }
}

// Listagem para o Admin do Sistema (paginada, filtrável).
async function list({ page = 1, limit = 25, action, q } = {}) {
  const take = Math.min(Math.max(1, Number(limit) || 25), 100);
  const current = Math.max(1, Number(page) || 1);
  const where = {};
  if (action) where.action = action;
  if (q) {
    where.OR = [
      { entityRef: { contains: q, mode: 'insensitive' } },
      { actorName: { contains: q, mode: 'insensitive' } },
    ];
  }
  const [total, items, actions] = await Promise.all([
    prisma.auditLog.count({ where }),
    prisma.auditLog.findMany({ where, orderBy: { createdAt: 'desc' }, skip: (current - 1) * take, take }),
    prisma.auditLog.groupBy({ by: ['action'], _count: { _all: true }, orderBy: { _count: { action: 'desc' } } }),
  ]);
  return {
    items, total, page: current, pages: Math.max(1, Math.ceil(total / take)),
    actions: actions.map((a) => ({ action: a.action, count: a._count._all })),
  };
}

module.exports = { actorFrom, anonimoFrom, contextoFrom, record, recordSafe, list };

File: backend-java/src/main/java/ao/kixima/audit/AuditService.java
package ao.kixima.audit;

import ao.kixima.audit.dto.AuditActionCount;
import ao.kixima.audit.dto.AuditLogDto;
import ao.kixima.audit.dto.AuditLogListResponse;
import ao.kixima.security.CurrentUser;
import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import jakarta.servlet.http.HttpServletRequest;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Sort;
import org.springframework.jdbc.core.JdbcTemplate;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.Instant;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.UUID;

/**
 * Espelha backend/src/services/auditService.js — trilho de auditoria
 * APPEND-ONLY: {@code actorFrom}/{@code anonimoFrom}/
 * {@code contextoFrom}/{@code recordSafe} (escrita, M1) e {@code list()}
 * (leitura paginada/filtrável para o Admin do Sistema, M5).
 */
@Service
public class AuditService {

    private static final Logger log = LoggerFactory.getLogger(AuditService.class);
    private static final int POR_OMISSAO = 25;
    private static final int MAXIMO = 100;

    private final AuditLogRepository repository;
    private final ObjectMapper objectMapper;
    private final JdbcTemplate jdbcTemplate;

    public AuditService(AuditLogRepository repository, ObjectMapper objectMapper, JdbcTemplate jdbcTemplate) {
        this.repository = repository;
        this.objectMapper = objectMapper;
        this.jdbcTemplate = jdbcTemplate;
    }

    public Actor actorFrom(CurrentUser user, HttpServletRequest req) {
        if (user == null) return anonimoFrom(req);
        return new Actor(user.id(), user.name(), user.role() == null ? null : user.role().name(),
                user.companyId(), req.getRemoteAddr());
    }

    public Actor anonimoFrom(HttpServletRequest req) {
        return Actor.anonimo(req.getRemoteAddr());
    }

    public Map<String, Object> contextoFrom(HttpServletRequest req) {
        Map<String, Object> m = new LinkedHashMap<>();
        m.put("ip", req.getRemoteAddr());
        String agente = req.getHeader("user-agent");
        m.put("agente", agente == null ? null : agente.substring(0, Math.min(180, agente.length())));
        return m;
    }

    public record Entry(Actor actor, String action, String entityType, String entityId,
                         String entityRef, Object detail) {
        public Entry(Actor actor, String action, String entityType) {
            this(actor, action, entityType, null, null, null);
        }
    }

    /**
     * Variante "não pode partir o negócio": regista após o sucesso da
     * operação; uma falha aqui vai para o log (Sentry entra quando M0
     * adicionar essa dependência — ver TODO em GlobalExceptionHandler).
     */
    public void recordSafe(Entry entry) {
        try {
            String detailJson = entry.detail() == null ? null : objectMapper.writeValueAsString(entry.detail());
            Actor a = entry.actor() == null ? Actor.anonimo(null) : entry.actor();
            AuditLog row = new AuditLog(UUID.randomUUID().toString(), entry.action(), entry.entityType(),
                    entry.entityId(), entry.entityRef(), a.actorId(), a.actorName(), a.actorRole(),
                    a.companyId(), a.ip(), detailJson, Instant.now());
            repository.save(row);
        } catch (Exception e) {
            log.error("AUDITORIA FALHOU — registo perdido: action={}", entry.action(), e);
        }
    }

    /** Listagem paginada/filtrável para o Admin do Sistema — GET /api/admin/audit-logs. */
    @Transactional(readOnly = true)
    public AuditLogListResponse list(Integer page, Integer limit, String action, String q) {
        int take = Math.min(Math.max(1, limit == null ? POR_OMISSAO : limit), MAXIMO);
        int current = Math.max(1, page == null ? 1 : page);

        Page<AuditLog> pagina = repository.findAll(AuditLogSpecifications.comFiltros(action, q),
                PageRequest.of(current - 1, take, Sort.by(Sort.Direction.DESC, "createdAt")));

        List<Map<String, Object>> contagens = jdbcTemplate.queryForList(
                "SELECT action, COUNT(*) AS total FROM audit_logs GROUP BY action ORDER BY total DESC");
        List<AuditActionCount> actions = contagens.stream()
                .map(r -> new AuditActionCount((String) r.get("action"), ((Number) r.get("total")).longValue()))
                .toList();

        return new AuditLogListResponse(pagina.getContent().stream().map(this::toDto).toList(),
                pagina.getTotalElements(), current, Math.max(1, pagina.getTotalPages()), actions);
    }

    private AuditLogDto toDto(AuditLog row) {
        JsonNode detail = null;
        if (row.getDetail() != null) {
            try {
                detail = objectMapper.readTree(row.getDetail());
            } catch (Exception ignorado) {
                // detail corrompido/não-JSON — fica null em vez de rebentar a listagem inteira.
            }
        }
        return new AuditLogDto(row.getId(), row.getAction(), row.getEntityType(), row.getEntityId(), row.getEntityRef(),
                row.getActorId(), row.getActorName(), row.getActorRole(), row.getCompanyId(), row.getIp(), detail, row.getCreatedAt());
    }
}
